tests_integration/test_snowpark_function.py

```diff
@@ -0,0 +1,417 @@
+from tests_integration.conftest import runner
+from tests_integration.snowflake_connector import create_database, snowflake_session
+from tests_integration.testing_utils.sql_utils import sql_test_helper
+from tests_integration.testing_utils.naming_utils import object_name_provider
+from tests_integration.testing_utils.working_directory_utils import (
+    temporary_working_directory,
+)
+
+from tests_integration.conftest import SnowCLIRunner
+from tests_integration.testing_utils.assertions.test_file_assertions import (
+    assert_that_current_working_directory_contains_only_following_files,
+    assert_that_file_content_is_equal_to_snapshot,
+)
+from tests_integration.testing_utils.assertions.test_result_assertions import (
+    assert_that_result_is_successful_and_has_no_output,
+    assert_that_result_is_successful_and_output_contains,
+    assert_that_result_is_successful,
+    assert_that_result_is_successful_and_json_output_contains_value_at_path,
+)
+from tests_integration.testing_utils.naming_utils import ObjectNameProvider
+from tests_integration.testing_utils.sql_utils import SqlTestHelper
+
+
+@pytest.mark.integration
+def test_snowpark_function_flow(_test_steps):
+    _test_steps.assert_that_no_functions_are_in_snowflake()
+    _test_steps.assert_that_no_files_are_staged_in_snowflake()
+
+    _test_steps.snowpark_function_list_should_return_no_data()
+
+    _test_steps.snowpark_function_init_should_initialize_files_with_default_content()
+    _test_steps.snowpark_function_package_should_zip_files()
+
+    function_name = _test_steps.snowpark_function_create_should_finish_successfully()
+    _test_steps.assert_that_only_these_functions_are_in_snowflake(
+        f"{function_name}() RETURN VARCHAR"
+    )
+    _test_steps.assert_that_only_these_files_are_staged_in_snowflake(
+        f"deployments/{function_name}/app.zip"
+    )
+
+    _test_steps.snowpark_function_list_should_return_list_containing_function_at_the_first_place(
+        function_name=function_name,
+        arguments="()",
+        result_type="VARCHAR",
+    )
+
+    _test_steps.snowpark_function_describe_should_return_function_description(
+        function_name=function_name,
+        arguments="()",
+    )
+
+    _test_steps.snowpark_function_execute_should_return_expected_value(
+        function_name=function_name,
+        arguments="()",
+        expected_value="Hello World!",
+    )
+
+    _test_steps.snowpark_function_update_should_finish_successfully(function_name)
+    _test_steps.assert_that_only_these_functions_are_in_snowflake(
+        f"{function_name}() RETURN NUMBER"
+    )
+    _test_steps.assert_that_only_these_files_are_staged_in_snowflake(
+        f"deployments/{function_name}/app.zip"
+    )
+
+    _test_steps.snowpark_function_list_should_return_list_containing_function_at_the_first_place(
+        function_name=function_name,
+        arguments="()",
+        result_type="NUMBER",
+    )
```

Comment:
I think one test `function list` is enough

Reply:
I described in **[this comment](https://github.com/Snowflake-Labs/snowcli/pull/245#discussion_r1273658415)** why there are few invocations of `function list`. But IMO we should talk about it. I understand it may cost us execution time.